mlir/lib/Dialect/MQTOpt/Transforms/Transpilation/Common.cpp

```diff
@@ -91,4 +94,81 @@ bool isTwoQubitGate(UnitaryInterface op) {
 }
+
+[[nodiscard]] SWAPOp createSwap(mlir::Location location, mlir::Value in0,
+                                mlir::Value in1,
+                                mlir::PatternRewriter& rewriter) {
+  const mlir::SmallVector<mlir::Type> resultTypes{in0.getType(), in1.getType()};
+  const mlir::SmallVector<mlir::Value> inQubits{in0, in1};
+
+  return rewriter.create<SWAPOp>(
+      /* location = */ location,
+      /* out_qubits = */ resultTypes,
+      /* pos_ctrl_out_qubits = */ mlir::TypeRange{},
+      /* neg_ctrl_out_qubits = */ mlir::TypeRange{},
+      /* static_params = */ nullptr,
+      /* params_mask = */ nullptr,
+      /* params = */ mlir::ValueRange{},
+      /* in_qubits = */ inQubits,
+      /* pos_ctrl_in_qubits = */ mlir::ValueRange{},
+      /* neg_ctrl_in_qubits = */ mlir::ValueRange{});
+}
+
+void replaceAllUsesInRegionAndChildrenExcept(mlir::Value oldValue,
+                                             mlir::Value newValue,
+                                             mlir::Region* region,
+                                             mlir::Operation* exceptOp,
+                                             mlir::PatternRewriter& rewriter) {
+  if (oldValue == newValue) {
+    return;
+  }
+
+  rewriter.replaceUsesWithIf(oldValue, newValue, [&](mlir::OpOperand& use) {
+    mlir::Operation* user = use.getOwner();
+    if (user == exceptOp) {
+      return false;
+    }
+
+    // For other blocks, check if in region tree
+    mlir::Region* userRegion = user->getParentRegion();
+    while (userRegion) {
+      if (userRegion == region) {
+        return true;
+      }
+      userRegion = userRegion->getParentRegion();
+    }
+    return false;
+  });
+}
+
+[[nodiscard]] bool isExecutable(UnitaryInterface op, const Layout& layout,
+                                const Architecture& arch) {
+  assert(isTwoQubitGate(op));
+  const auto ins = getIns(op);
+  return arch.areAdjacent(layout.lookupHardwareIndex(ins.first),
+                          layout.lookupHardwareIndex(ins.second));
+}
+
+void insertSWAPs(mlir::Location loc, mlir::ArrayRef<QubitIndexPair> swaps,
+                 Layout& layout, mlir::PatternRewriter& rewriter) {
+  for (const auto [hw0, hw1] : swaps) {
+    const mlir::Value in0 = layout.lookupHardwareValue(hw0);
+    const mlir::Value in1 = layout.lookupHardwareValue(hw1);
+    [[maybe_unused]] const auto [prog0, prog1] =
+        layout.getProgramIndices(hw0, hw1);
+
+    auto swap = createSwap(loc, in0, in1, rewriter);
+    const auto [out0, out1] = getOuts(swap);
+
+    rewriter.setInsertionPointAfter(swap);
+    replaceAllUsesInRegionAndChildrenExcept(in0, out1, swap->getParentRegion(),
+                                            swap, rewriter);
+    replaceAllUsesInRegionAndChildrenExcept(in1, out0, swap->getParentRegion(),
+                                            swap, rewriter);
+
+    layout.swap(in0, in1);
+    layout.remapQubitValue(in0, out0);
+    layout.remapQubitValue(in1, out1);
```

Comment:
couldn't this be handled via the `layout.remap(swap)` method that will potentially be introduced?